block-layer/codecs/dag-jose.md

```diff
@@ -0,0 +1,100 @@
+
+Below we present an IPLD schema representing the encoded JOSE objects. Note that the `EncodedJOSE` union is not in fact a valid IPLD schema as there is no valid discriminator. The actual wire format is a single struct which contains all the keys from both the `EncodedJWE` and the `EncodedJWS` structs, implementors should follow [section 9 of the JWE spec](https://tools.ietf.org/html/rfc7516#section-9) and distinguish between these two branches of the union by checking if the `payload` attribute exists, and hence you have a JWS; or the `ciphertext` attribute, hence you have a JWE.
+
+**Encoded JOSE**
+
+```ipldsch
+type EncodedSignature struct {
+  header optional {String:Any}
+  protected optional Bytes
+  signature Bytes
+}
+
+type EncodedRecipient struct {
+  encrypted_key optional Bytes
+  header optional {String:Any}
+}
+
+type EncodedJWE struct {
+  aad optional Bytes
+  ciphertext Bytes
+  iv optional Bytes
+  protected optional Bytes
+  recipients [EncodedRecipient]
+  tag optional Bytes
+  unprotected optional {String:Any}
+}
+
+type EncodedJWS struct {
+  signatures [EncodedSignature]
+  payload optional Bytes
+}
+
+type EncodedJOSE union { EncodedJWE | EncodedJWS }
+```
+
+## Padding for encryption
+
+Applications may need to pad the plaintext when encrypting to avoid leaking the size of the plaintext. This raises the question of how the application knows what part of the decrypted plaintext is padding. In this case we use the fact that the plaintext MUST be a valid CID, implementations should parse the plaintext as a CID and discard any content beyond the multihash digest size - which we assume to be the padding.
+
+
+## Decoded JOSE
+
+Typically implementations will want to decode this format into something more useful for applications. Exactly what that will look like depends on the language of the implementation, here we use the IPLD schema language to give a somewhat language agnostic description of what the decoded representation might look like at runtime. Note that everything which is specified as `base64url(ascii(<some JSON>))` in the JOSE specs - and which we encode as `Bytes` in the wire format - is here decoded to a `String`. We also add the `link: &Any` attribute to the `DecodedJWS`,  which allows applications to easily retrieve the authenticated content.
+
+Also note that - as with the encoded representation - the `DecodedJOSE` union is not valid IPLD schema as there is no way to discriminate between them. How exactly this would be represented will depend on the language of the implementation. For example in Typescript this would be a straightforward `type DecodedJOSE = DecodedJWE | DecodedJWS`.
+
+```ipldsch
+type DecodedSignature struct {
+  header optional {String:Any}
+  protected optional String
+  signature String
+}
+
+type DecodedJWS struct {
+  payload String
+  signatures [DecodedSignature]
+  link: &Any
```

Comment:
@warpfork any preference here on using `&Any` vs `Link` since they are both effectively the same thing.

Reply:
I think I tend towards likely `&Any` but it's true that either is acceptable.  I don't think any of our tools make a functional distinction at all, as things presently stand.

Reply:
there's `:` on this line which is a syntax error, I don't suppose someone in this issue wants to PR a fix?